common/src/collision.rs

```diff
@@ -0,0 +1,129 @@
+
+    const EPSILON: f32 = 1e-4;
+    let found_collision = if node_ray.direction.z >= 0.0 {
+        // Going the wrong way. Don't report a collision.
+        false
+    } else if (-EPSILON..=0.0).contains(&node_ray.position.z) {
+        // Extra logic to ensure precision issues don't allow a collider to clip through a surface
+        endpoint.tanh_distance = 0.0;
+        true
+    } else {
+        // Compute needed tanh_distance to reach z=0 plane.
+        let tanh_distance = -node_ray.position.z / node_ray.direction.z;
+        if tanh_distance >= 0.0 && tanh_distance < endpoint.tanh_distance {
+            // Report a collision if the needed tanh_distance is in the right range
+            endpoint.tanh_distance = tanh_distance;
+            true
+        } else {
+            false
+        }
+    };
+
+    if found_collision {
+        endpoint.hit = Some(RayHit {
+            normal: math::mtranspose(&start_node_transform) * na::Vector4::z(),
+        });
+    }
+
+    endpoint
+}
+
+/// Where a cast ray ended, and all information about the hit at the end if such a hit occurred.
+#[derive(Debug)]
+pub struct RayEndpoint {
+    /// The tanh of the length of the resulting ray segment so far. As new intersections are found, the
+    /// ray segment gets shorter each time.
+    pub tanh_distance: f32,
+
+    /// Information about the intersection at the end of the ray segment. If this is `None`, there
+    /// are no intersections.
+    pub hit: Option<RayHit>,
+}
+
+/// Information about the intersection at the end of a ray segment.
+#[derive(Debug)]
+pub struct RayHit {
+    /// Represents the normal vector of the hit surface in the original coordinate system
+    /// of the ray tracing. To get the actual normal vector, project it so that it is orthogonal
+    /// to the ray's endpoint in Lorentz space.
+    pub normal: na::Vector4<f32>,
+}
+
+/// A ray in hyperbolic space. The fields must be lorentz normalized, with `mip(position, position) == -1`,
+/// `mip(direction, direction) == 1`, and `mip(position, direction) == 0`.
+#[derive(Debug)]
+pub struct Ray {
+    pub position: na::Vector4<f32>,
+    pub direction: na::Vector4<f32>,
+}
+
+impl Ray {
+    pub fn new(position: na::Vector4<f32>, direction: na::Vector4<f32>) -> Ray {
+        Ray {
+            position,
+            direction,
+        }
+    }
+
+    /// Returns a point along this ray `atanh(tanh_distance)` units away from the origin. This point
+    /// is _not_ lorentz normalized.
+    pub fn ray_point(&self, tanh_distance: f32) -> na::Vector4<f32> {
+        self.position + self.direction * tanh_distance
+    }
+}
+
+impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
+    type Output = Ray;
+
+    #[inline]
+    fn mul(self, rhs: &Ray) -> Self::Output {
+        Ray {
+            position: self * rhs.position,
+            direction: self * rhs.direction,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
This is a good start, but we should always also test *negative* cases. For example, a ray parallel to the plane (though I guess that's less interesting in H3, heh) and a ray pointing away from it.

Reply:
Oh yeah. I didn't try that hard to add extensive tests here because the code was going to be deleted.